presto-main/src/main/java/com/facebook/presto/operator/HttpPageBufferClient.java

```diff
@@ -343,6 +366,16 @@ public void onSuccess(PagesResponse result)
 
                         if (token == HttpPageBufferClient.this.token) {
                             pages = result.getPages();
+                            if (pages.size() > pendingPages.size()) {
```

Comment:
This means we have fetched more pages than what we know. For example, suppose a client has 3 pending pages, then the exchange client should only give the client 3 pending pages worth of quota even if the server may have 5 pages. If the client comes back with 5 pages, there is definitely something wrong with the scheduler.